stdlib/nats/src/main/java/org/ballerinalang/nats/nativeimpl/producer/RequestReply.java

```diff
@@ -0,0 +1,93 @@
+import org.ballerinalang.bre.bvm.CallableUnitCallback;
+import org.ballerinalang.connector.api.BLangConnectorSPIUtil;
+import org.ballerinalang.connector.api.Struct;
+import org.ballerinalang.model.NativeCallableUnit;
+import org.ballerinalang.model.types.TypeKind;
+import org.ballerinalang.model.values.BMap;
+import org.ballerinalang.model.values.BString;
+import org.ballerinalang.model.values.BValue;
+import org.ballerinalang.model.values.BValueArray;
+import org.ballerinalang.natives.annotations.BallerinaFunction;
+import org.ballerinalang.natives.annotations.Receiver;
+import org.ballerinalang.nats.nativeimpl.Constants;
+import org.ballerinalang.nats.nativeimpl.Utils;
+
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Sends message to a given subject and receive a response.
+ *
+ * @since 0.995
+ */
+@BallerinaFunction(
+        orgName = "ballerina",
+        packageName = "nats",
+        functionName = "sendRequestReplyMsg",
+        receiver = @Receiver(type = TypeKind.OBJECT, structType = "Producer", structPackage = "ballerina/nats"),
+        isPublic = true
+)
+public class RequestReply implements NativeCallableUnit {
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void execute(Context context, CallableUnitCallback callback) {
+        try {
+            Struct publisher = Utils.getReceiverObject(context);
+            String subject = context.getStringArgument(0);
+            byte[] content = ((BValueArray) context.getRefArgument(1)).getBytes();
+            StreamingConnection connection = (StreamingConnection) publisher.getStructField(Constants.CONNECTION_STRUCT)
+                    .getNativeData(Constants.NATS_CONNECTION);
+            CompletableFuture<Message> future = connection.getNatsConnection().request(subject, content);
+            // this will be a blocking call, given we will be rewriting this in ballerina
+            // will not invest to write java non blocking version
+            Message message = future.get();
+            BMap<String, BValue> msgObj = BLangConnectorSPIUtil.createBStruct(context,
+                    Constants.NATS_PACKAGE,
+                    Constants.NATS_MESSAGE_OBJ_NAME);
+            msgObj.addNativeData(Constants.NATS_MSG, message);
+            msgObj.put(Constants.MSG_CONTENT_NAME, new BString(new String(message.getData(), StandardCharsets.UTF_8)));
+            context.setReturnValues(msgObj);
+        } catch (InterruptedException ignore) {
+            Thread.currentThread().interrupt();
+        } catch (ExecutionException e) {
+            context.setReturnValues(Utils.createError(context, Constants.NATS_ERROR_CODE, e.getMessage()));
+        } finally {
+            callback.notifySuccess();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isBlocking() {
+        return false;
```

Comment:
I guess this function can be blocking if there are no particular reason for making it non-blocking.

Reply:
We use the callback to return the values here, we would gradually introduce the separate executor to handle this asynchronously which would be addressed in future.